runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
+     * execute another quantum program. Configurable execution parameters, whether explicit in
+     * the interface (like shots) or internal via the constructor kwargs, should be maintained
+     * in the same state as when entering this method.
+     */
+    virtual void ReleaseAllQubits() = 0;
+
+    /**
+     * @brief Get the number of currently allocated qubits.
      *
-     * @param qubit The id of the qubit
+     * @return `size_t`
      */
-    virtual void ReleaseQubit(QubitIdType qubit) = 0;
+    virtual auto GetNumQubits() const -> size_t = 0;
 
     /**
-     * @brief Release all qubits.
+     * @brief (Optional) Allocate a qubit.
+     *
+     * Allocate a new qubit in the |0> state. This method is only needed for devices with dynamic
+     * qubit allocation.
+     *
+     * See `AllocateQubits` for more details on allocation semantics.
+     *
+     * @return `QubitIdType` Logical qubit ID.
      */
-    virtual void ReleaseAllQubits() = 0;
+    virtual auto AllocateQubit() -> QubitIdType
+    {
+        RT_FAIL("Dynamic qubit allocation is unsupported by device");
+    }
 
     /**
-     * @brief Get the number of allocated qubits.
+     * @brief (Optional) Release a qubit.
      *
-     * @return `size_t`
+     * Release the provided qubit. This method is only needed for devices with dynamic qubit
+     * allocation.
+     *
+     * Note that the interface does not require the qubit to be in a particular state. The behaviour
+     * for releasing an entangled qubit is left up to the device, and could include:
+     *  - raising a runtime error (assuming the device can detect impure states)
+     *  - continuing execution with an entangled but inaccessible qubit
+     *  - resetting the qubit with or without measuring its state
+     *
+     * Opposite of `AllocateQubit`.
+     *
+     * @param qubit ID of the qubit to release.
      */
-    [[nodiscard]] virtual auto GetNumQubits() const -> size_t = 0;
+    virtual void ReleaseQubit([[maybe_unused]] QubitIdType qubit)
+    {
+        RT_FAIL("Dynamic qubit release is unsupported by device");
+    }
+
+    // ----------------------------------------
+    //  EXECUTION MANAGEMENT
+    // ----------------------------------------
 
     /**
-     * @brief Set the number of device shots.
+     * @brief Set the number of execution shots.
+     *
+     * The Runtime will call this function to set the number of times the quantum execution is to
+     * be repeated. Generally, it will be called once before any quantum instructions are run, but
+     * some devices may choose to support it at arbitrary points in the program if they have the
+     * capability to simulate shot noise.
      *
-     * @param shots The number of noise shots
+     * Devices with no or restricted support are encouraged to raise a runtime error.
+     *
+     * @param shots Shot number.
      */
     virtual void SetDeviceShots(size_t shots) = 0;
```

Comment:
This method as well as `GetDeviceShots` could be optional too. What do you think?

Reply:
Erick asked a similar question, given how fundamental shots are and that we don't have a single device not supporting shots, I'm inclined to leave it.